index.js

```diff
@@ -161,178 +88,249 @@ function invalidationsAsPatches(invalidated, currentTree, nextTree) {
   return patches;
 }
 
-Filter.prototype.build = function() {
-  // @ts-ignore
-  let srcDir = this.inputPaths[0];
-  // @ts-ignore
-  let destDir = this.outputPath;
+function invoke(context, fn, args) {
+  return new Promise(resolve => {
+    resolve(fn.apply(context, args));
+  });
+}
+
+class Filter extends Plugin {
+  static shouldPersist(env, persist) {
+    let result;
+
+    if (env.CI) {
+      result = persist && env.FORCE_PERSISTENCE_IN_CI;
+    } else {
+      result = persist;
+    }
 
-  if (this.dependencyInvalidation && !this.dependencies) {
-    this.dependencies = this.processor.initialDependencies(srcDir);
+    return !!result;
   }
 
-  if (this._needsReset) {
+  constructor (inputTree, options) {
+    super([inputTree], {
+      name: (options && options.name),
+      annotation: (options && options.annotation),
+      persistentOutput: true
+    });
+    if (!(this instanceof Filter) || Object.getPrototypeOf(this) === Filter.prototype) {
+      throw new TypeError('Filter is an abstract class and must be sub-classed');
```

Comment:
If we keep this error, let's provide a clearer error message. 

1. more specific name: `Filter` -> `BroccoliPersistentFilter`
2. slightly clearer (i think) error message: `BroccoliPersistentFilter Is an abstract class, and cannot be instantiated directly, rather is intended to be sub-classed`